### ISSUE TITLE
PHP strict typehinting or in docblocks?

### ISSUE THREAD

**Issue by cbl** [aw-studio/guidelines#19]

**Comment by jannescb** [comment]
> As @cbl said in our conversation, this is exactly why you should not use type declarations in a reused open source package, as it makes it difficult to extend the package.
> 
> In other, closed sourced / client projects they should be used whenever possible, I would say.

I second that! 

Use, whenever it's possible.

**Comment by jannescb** [comment]
I also think there is nothing wrong using both. When a function is fully described (input and output types) the auto-generator for the docblock will take care of it.

**Comment by lpheller** [comment]
Well, you could argue that adding a Dockblock is mainly unnesseary, when using type hints, return declaration and a good method name...

```php
public function getFirstUrlSegment(string $url): string {
    // ... 
}
```

Adding a dockblock mainly adds visual noise without providing any extra information.
```php
/**
 * Get the first url segment from a URL string.
 *
 * @params string $url
 * @return string
 */
public function getFirstUrlSegment(string $url): string {
    // ... 
}
```

But I personally prefer that noise to the lack of a dock block with description. I actually find it rather more annoying when some methods have a dockblock and others don't. Either all of them or none at all, please. 🤓

**Comment by jannescb** [comment]
This is a typical case of "it depends".

But it might be a good goal for closed source / customer projects to get rid of docblocks if a function is not too complex by being as meaningful in naming as possible.

Your example doesn't need it, it is perfectly clear.

Personally, I like the physical distance between the functions created by the blocks, but this is just a matter of getting used to the typographic character of the program 😆

**Comment by cbl** [comment]
Sometimes typehinting results in objects being formatted. One must keep in mind that this might not be wanted. Given the following example, a class that can be formatted to a string:

```php
class SomeClass
{
    public $value;

    public function __toString();
    {
        return $value;
    }
}

class UsingTypeHint
{
    public $value;

    public function setValue(string $value)
    {
        $this->value = $value;
    }
}

class WithoutTypeHint
{
    public $value;

    public function setValue($value)
    {
        $this->value = $value;
    }
}
```

Changing the value of `SomeClass` after it was passed to `setValue` will end up in different results:

```php
$class = new SomeClass;
$usingTypeHint = new UsingTypeHint;
$withoutTypeHint = new WithoutTypeHint;

$class->value = 'foo';
$usingTypeHint->setValue($class);
$withoutTypeHint->setValue($class);
$class->value = 'bar';

echo $usingTypeHint->value; // will be "foo"
echo $withoutTypeHint->value; // will be "bar"
```

**Comment by lpheller** [comment]
>Personally, I like the physical distance between the functions created by the blocks [...]
Exactly this! 😂

**Comment by cbl** [comment]
@lpheller No docblocks at all is no option in my view 🤔

**Comment by jannescb** [comment]
> @lpheller No docblocks at all is no option in my view 🤔

I think spatie does it that way, too…

**Comment by jannescb** [comment]
>  I agree that docblocks are unnecessary for closed source projects when the function name describes what the function does, however sometimes its hard to find good names for a function.

Yes! If you can't describe what the function does in a short sentence use a docblock.

**Comment by lpheller** [comment]
@cbl No, of course not!  I didn't meant to say that in any case. This was only about to the "visual appearance" of a php class, that it looks nicer when either all methods are commented or none, imho. 

I'd personally prefer more dockblocks than none.

@jannescb Spatie does this actually pretty inconsistent. Looking at the media library e.g. some files have no dockblock at all, some have all methods commented and some only a few... 
Looking at this right now, I'd definitely say without any dockblock doesn't look much nicer or more clear at all.

**Comment by cbl** [comment]
![EyYxG2aXAAEmKOr](https://user-images.githubusercontent.com/29352871/119666122-5abfde80-be35-11eb-83f0-74c56f414720.jpeg)

**Comment by jannescb** [comment]
$b
